File: coffee-system/src/main/java/com/jesperancinha/coffee/system/api/manager/CoffeeProcessor.java
package com.jesperancinha.coffee.system.api.manager;

import com.jesperancinha.coffee.system.api.concurrency.QueueCallable;
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.Coffees.Coffee;
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.PaymentTypes.Payment;
import org.jesperancinha.coffee.system.input.Employees.Employee;
import org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PostAction;

import java.util.List;

/**
 * Created by joao on 29-4-16.
 */
public interface CoffeeProcessor extends Processor {
    void callMakeCoffee(Employee employee, String name, Coffee chosenCoffee, Payment payment, List<PostAction> postActions, QueueCallable parentCallable);
}

File: jeorg-kotlin-apps/coffee-paradigms/coffee-system/src/main/kotlin/org/jesperancinha/coffee/system/manager/CoffeeProcessor.kt
package org.jesperancinha.coffee.system.manager

import org.jesperancinha.coffee.system.api.concurrency.QueueCallable
import org.jesperancinha.coffee.system.concurrency.CoffeeMainCallable
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeeMachine.Coffees.Coffee
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeeMachine.PaymentTypes.Payment
import org.jesperancinha.coffee.system.input.Employees.Employee
import org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PostAction
import org.jesperancinha.coffee.system.queues.Queue
import org.jesperancinha.coffee.system.queues.QueueCoffee
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.concurrent.Callable

@Service
class CoffeeProcessor(
    @Autowired
    private val queueCoffee: QueueCoffee,
    @Autowired
    private val paymentProcessor: PaymentProcessor
) : ProcessorAbstract() {

    fun callMakeCoffee(
        employee: Employee,
        name: String,
        coffee: Coffee,
        payment: Payment,
        postActions: List<PostAction>,
        parentCallable: QueueCallable?
    ) {
        val coffeeMainCallable = CoffeeMainCallable(employee, name, coffee, payment, postActions, paymentProcessor)
        parentCallable?.allCallables?.add(coffeeMainCallable)
    }

    override val executorServiceQueue: Queue = queueCoffee

    override fun getExecutorName(callable: Callable<Boolean>): String = (callable as CoffeeMainCallable).name

    fun addQueueSize(queueSize: Int, name: String) {
        queueCoffee.setQueueSize(queueSize, name)
    }

    fun initExecutors() {
        queueCoffee.initExecutors()
    }

    fun stopExecutors() {
        queueCoffee.stopExecutors()
    }
}